Sources/Apollo/ApolloClient.swift

```diff
@@ -28,7 +28,7 @@ public typealias OperationResultHandler<Operation: GraphQLOperation> = (_ result
 
 /// The `ApolloClient` class provides the core API for Apollo. This API provides methods to fetch and watch queries, and to perform mutations.
 public class ApolloClient {
-  let networkTransport: NetworkTransport
+  let link: TerminatingLink
```

Comment:
I agree, it could be confusing, but actually it makes quite a lot of sense.
The confusion comes from the belief that there is a chain of links, but there isn't.
The only way to combine links is by composition, which produces a new link.

Consider concatenation:
`concat(NT, T) -> T` and `concat(NT, NT) -> NT` where `NT` is a non-terminating link and `T` is a terminating link.

And split:
`split(NT, NT) -> NT` and `split(T, T) -> T`

Then there is also forced termination:
`makeTerminating(NT) -> T`

As you can see, no matter how many links go into the composition the end result is always a single link.

It makes sense to give the client a terminating link, because it is what produces a result in the end.
Note that this very terminating link may be a single link (e.g. `HTTPLink`) or a link that was composed from (multiple) other links (e.g. `concat(AuthLink, split(HTTPLink, WebSocketLink))`).

Does that make sense to you?

Reply:
Ah, you're right. I was thinking of this as the first link in the chain, but it's function composition. If we can get this working reliably that would be great, but I don't think it's absolutely necessary to make termination part of the type system. 

> However, it makes it difficult to define a split of a terminating and a non-terminating link or vice-versa, because what should the return type be? SchrödingersLink?

Have you thought more about this yet?

And what does `makeTerminating` do?